[Week1] Todo REST api 만들기

goodcodings opened the pull request:
1. DIP를 최대한 지키기 위해 역할과 구현을 나눴습니다
- 의존 주입을하는 AppConfig 클래스를 생성했습니다.
2. 에릭 에반스의 ddd sample을 참조하여 패키지 구조를 잡았습니다.
3. task의 빌더 패턴을 다시 만들었고 title 변경시 task 객체의 자율성을 위해 updateTitle 메소드를 생성했습니다
4. httpHandler를 구현하여 클라이언트로 부터 요청 받고 응답하는 코드를 생성했습니다
5. Todo 생성까지 구현했습니다
6. 각 메소드에 맞는 처리를 하기 위해 post,get,put,delete클래스를 생성했습니다.

goodcodings committed:
task 도메인생성

Signed-off-by: Lee Byoung Seong <dev.bslee@gmail.com>

goodcodings committed:
[Week1] 빌더패턴 제거

goodcodings committed:
[Week1] 빌더패턴 제거

goodcodings committed:
[Week1] Todo REST api 만들기
1. DIP를 최대한 지키기 위해 역할과 나눴습니다
   - 의존 주입을하는 AppConfig 클래스를 생성했습니다.
2. 에릭 에반스의 ddd sample을 참조하여 패키지 구조를 잡았습니다.
3. task의 빌더 패턴을 다시 만들었고 title 변경시 task 객체의 자율성을 위해 updateTitle 메소드를  생성했습니다

goodcodings committed:
Merge pull request #1 from eggNnuts/eggNnuts

[Week1] Todo REST api 만들기

Lee-Byoung-Seong committed:
[Week1] Todo REST api 만들기
1. DIP를 최대한 지키기 위해 역할과 구현을 나눴습니다
- 의존 주입을하는 AppConfig 클래스를 생성했습니다.
2. 에릭 에반스의 ddd sample을 참조하여 패키지 구조를 잡았습니다.
3. task의 빌더 패턴을 다시 만들었고 title 변경시 task 객체의 자율성을 위해 updateTitle 메소드를 생성했습니다
4. httpHandler를 구현하여 클라이언트로 부터 요청 받고 응답하는 코드를 생성했습니다
5. Todo 생성까지 구현했습니다
6. 각 메소드에 맞는 처리를 하기 위해 post,get,put,delete클래스를 생성했습니다.

goodcodings committed:
Merge pull request #2 from eggNnuts/eggNnuts

[Week1] Todo REST api 만들기

Lee-Byoung-Seong committed:
[Week1] Todo REST api 만들기
1. DIP를 최대한 지키기 위해 역할과 구현을 나눴습니다
- 의존 주입을하는 AppConfig 클래스를 생성했습니다.
2. 에릭 에반스의 ddd sample을 참조하여 패키지 구조를 잡았습니다.
3. task의 빌더 패턴을 다시 만들었고 title 변경시 task 객체의 자율성을 위해 updateTitle 메소드를 생성했습니다
4. httpHandler를 구현하여 클라이언트로 부터 요청 받고 응답하는 코드를 생성했습니다
5. Todo 생성까지 구현했습니다
6. 각 메소드에 맞는 처리를 하기 위해 post,get,put,delete클래스를 생성했습니다.

goodcodings committed:
Merge pull request #3 from eggNnuts/eggNnuts

[Week1] Todo REST api 만들기

Lee-Byoung-Seong committed:
[Week1] Todo REST api 만들기
1. DIP를 최대한 지키기 위해 역할과 구현을 나눴습니다
- 의존 주입을하는 AppConfig 클래스를 생성했습니다.
2. 에릭 에반스의 ddd sample을 참조하여 패키지 구조를 잡았습니다.
3. task의 빌더 패턴을 다시 만들었고 title 변경시 task 객체의 자율성을 위해 updateTitle 메소드를 생성했습니다
4. httpHandler를 구현하여 클라이언트로 부터 요청 받고 응답하는 코드를 생성했습니다
5. Todo 생성까지 구현했습니다
6. 각 메소드에 맞는 처리를 하기 위해 post,get,put,delete클래스를 생성했습니다.

goodcodings committed:
Merge pull request #4 from eggNnuts/eggNnuts

[Week1] Todo REST api 만들기

goodcodings committed:
Revert "[Week1] Todo REST api 만들기"

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/config/AppConfig.java:
```diff
@@ -0,0 +1,16 @@
+import com.codesoom.assignment.infrastructure.todo.TodoServiceImpl;
+
+public class AppConfig {
+    public final static int PORT = 1238;
+    public final static int BACK_LOG = 0;
+    public final static String HOST="localhost";
```
아마 다음에 추가하실 것 같지만 그래도 의견을 드려봅니다!

`static` 을 지우고 생성자에서 값을 받도록 하는 것을 어떨까요?

goodcodings replied:
네 내일 코드 올릴때 static은 지우고 생성자로 값을 받도록 수정 하겠습니다ㅎㅎ

ssisksl77 commented on app/build.gradle:
```diff
@@ -25,6 +25,8 @@ dependencies {
     // This dependency is used by the application.
     implementation 'com.google.guava:guava:29.0-jre'
+    // apache lang
+    implementation 'org.apache.commons:commons-lang3:3.0'
```
아주 좋아요! 이제 실행할 때 의존성 에러는 나지 않네요!
PR은 기능은 미완이더라도 실행이 되는 것이 좋아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/server/TodoServerHandler.java:
```diff
@@ -0,0 +1,80 @@
+        return httpMethodToConstructorMap;
+    }
+
+    private void writeHttpResponse(HttpExchange httpExchange, HttpResponse httpResponse) throws IOException {
+        httpExchange.sendResponseHeaders(httpResponse.getStatusCode(), httpResponse.getBody().getBytes().length);
+        OutputStream outputStream = httpExchange.getResponseBody();
+        outputStream.write(httpResponse.getBody().getBytes());
+        outputStream.flush();
+        outputStream.close();
+    }
```
```suggestion
    private void writeHttpResponse(HttpExchange httpExchange, HttpResponse httpResponse) {
        try (OutputStream outputStream = httpExchange.getResponseBody()) {
            var body = httpResponse.getBody().getBytes();
            httpExchange.sendResponseHeaders(httpResponse.getStatusCode(), body.length);
            outputStream.write(body);
            outputStream.flush();
        } catch (IOException e){
            System.out.printf("in handle -> {}", e.getMessage());
        }
    }
```

한 번 다음처럼 바꿔보았어요!
try-catch 문이 너무 크게 잡혀있으면 어느 부분 때문에 `IOException` 을 잡아야 하는지 알기 힘들 수가 있어요!
또한 `OutputStream` 은 `Closeable` 을 구현하고 있어서 `try-with-resources statement` 를 이용하면 예외가 나건 안나건 자동으로 닫아져요! 

기존 코드는 중간에 예외가 나면 잘 닫히지 않을 수 있어요!

goodcodings replied:
감사합니다!!
IO  스트림 관련 클래스의 예외처리할때는 try-with-resources statement를 사용해서 자원 반납을 자동으로 하고 

예외를 잡을때는 큰 범위가 아닌 꼭 필요한 범위를 잡도록 하겠습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,25 @@
+
+    public void runTodoServer() {
+        try {
+            AppConfig appConfig =new AppConfig();
```
```suggestion
            AppConfig appConfig = new AppConfig();
```

너~무 사소하지만 `=` 연산자를 쓸 때 코딩 포맷이 조금씩 다른 것 같아요!
어떻게 해도 상관은 없겠지만 일단은 일관되게 하는 것이 중요해요!
바로 위 코드처럼 양쪽 모두 스페이스를 붙이는 것이 좋을 것 같아요.

goodcodings commented:
안녕하세요 과제를 수행하면서 다른 분들의 작성하신 코드를 참조하고 있는데 다른분들은 todo id를 시퀀스로 하고 계셔서....  
제가 잘못하고 있는게 아닐까해서 질문을 드립니다...
객체의 자율성을 보장 할려고 객체 내부에서 아이디를 생성 하는 의미로 코딩을 했는데...

제가 잘못된 방향으로 과제를 수행하고 있으면 바로 수정하겠습니다...

goodcodings commented:
## 요구사항
 - [ ] **ToDo 목록 조회하기** 
-  메소드 : GET
-  경로 :  /tasks

 - [ ] **ToDo 상세 조회하기** 
-  메소드 : GET
-  경로 :  /tasks/{id}
-  예외처리 고민
     - pathvariable가 존재하지 않을 경우 

 - [ ] **ToDo 생성하기** 
 -  메소드 : POST
 -  경로 : /tasks
 - 예외처리 고민
     - title이 존재 하지 않거나 
     - 추가 고민으로 문자만 받을수 있게 제한해야 할까

- [ ] **ToDo 수정하기** 
 -  메소드 : PUT/PATCH 
 -  경로 : /tasks/{id}
 - 예외처리 고민
     - pathvariable가 존재하지 않을 경우 
     - 수정할려는 task가 존재하지 않을 경우 어떻게 처리할 것인지
     - title이 존재 하지 않거나 문자만 받을수 있게 제한해야 할까

- [ ] **ToDo 삭제하기** 
 -  메소드 : DELETE 
 -  경로 : /tasks/{id}
 - 예외처리 고민
     - pathvariable가 존재하지 않을 경우 
     - 수정할려는 task가 존재하지 않을 경우 어떻게 처리할 것인지

goodcodings committed:
Update app/src/main/java/com/codesoom/assignment/server/TodoServerHandler.java

Co-authored-by: 남영환 <ssisksl77@gmail.com>

Lee-Byoung-Seong committed:
[Week1] Todo REST api 만들기
1.feat/ 로거 클래스  생성

Lee-Byoung-Seong committed:
Merge remote-tracking branch 'upstream/revert-4-eggNnuts' into eggNnuts

# Conflicts:
#	app/src/main/java/com/codesoom/assignment/config/AppConfig.java
#	app/src/main/java/com/codesoom/assignment/server/TodoServerHandler.java

goodcodings committed:
Merge pull request #5 from eggNnuts/eggNnuts

[Week1] Todo REST api ¸¸µé±â

goodcodings committed:
Merge branch 'main' into revert-4-eggNnuts

goodcodings committed:
Merge pull request #6 from eggNnuts/revert-4-eggNnuts

feat 로그 클래스 생성

goodcodings committed:
[Week1] Todo REST api 만들기
[수행 사항]
1. method, httpStatusCode enum 객체 생성했습니다.
2. HttpResponse 객체에 Serializable 추가했습니다.
3. HttpResponse 객체를 bype array로 만들기 위해 유틸 함수 생성했습니다
4. pathVariable 동적으로 얻기 위한 메소드를 생성했습니다

[내일 수행사항]
1. 응답시 body에 데이터가 실리지 않는 원인 분석

goodcodings committed:
[Week1] Todo REST api 만들기
[수행 사항]
1. method, httpStatusCode enum 객체 생성했습니다.
2. HttpResponse 객체에 Serializable 추가했습니다.
3. HttpResponse 객체를 bype array로 만들기 위해 유틸 함수 생성했습니다
4. pathVariable 동적으로 얻기 위한 메소드를 생성했습니다

[내일 수행사항]
1. 응답시 body에 데이터가 실리지 않는 원인 분석

goodcodings committed:
Merge remote-tracking branch 'upstream/revert-4-eggNnuts' into eggNnuts

# Conflicts:
#	app/src/main/java/com/codesoom/assignment/server/TodoServerHandler.java

goodcodings committed:
Merge pull request #7 from eggNnuts/main

Merge pull request #6 from eggNnuts/revert-4-eggNnuts

goodcodings committed:
[Week1] Todo REST api 만들기
[수행 사항]
1. method, httpStatusCode enum 객체 생성했습니다.
2. HttpResponse 객체에 Serializable 추가했습니다.
3. HttpResponse 객체를 bype array로 만들기 위해 유틸 함수 생성했습니다
4. pathVariable 동적으로 얻기 위한 메소드를 생성했습니다

[내일 수행사항]
1. 응답시 body에 데이터가 실리지 않는 원인 분석

goodcodings committed:
Merge remote-tracking branch 'upstream/eggNnuts' into eggNnuts

# Conflicts:
#	app/src/main/java/com/codesoom/assignment/common/request/HttpMethod.java
#	app/src/main/java/com/codesoom/assignment/common/request/HttpRequest.java
#	app/src/main/java/com/codesoom/assignment/common/response/HttpResponse.java
#	app/src/main/java/com/codesoom/assignment/common/response/HttpResponseCode.java
#	app/src/main/java/com/codesoom/assignment/infrastructure/mapping/PostHttpMapping.java

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/infrastructure/repository/MemoryRepository.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    public List<Todo> allTodos() {
+        return store.values().stream().collect(Collectors.toList());
```
```suggestion
        new ArrayList<>(store.values());
```

스트림을 만들고 그것을 다시 `List` 로 변환하는 것보다 바로 List로 만드는 것이 효율적일 것 같아요!

goodcodings replied:
엇...... 이런 방법이 있었군요 ..... 감사합니다!! 바로 반영하겠습니다 ㅎㅎ

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/domain/repository/Repository.java:
```diff
@@ -0,0 +1,14 @@
+public interface Repository {
+
+    Todo save(Todo todoId);
+    Optional<Todo> findById(String todoId);
+    Todo getById(String todoId);
```
안녕하세요! 
`findById` 와 `getById` 는 어떻게 나뉘어서 사용하기 위해 분리하신 걸까요?!

goodcodings replied:
용도별로 사용하기 위해서 findById와 getById를 나눴습니다...ㅎㅎㅎ

단순 조회를 할때는 찾을려는 ToDo가 없더라도 빈 값을 return해도 되지만 
업데이트와 삭제의 경우 찾는 ToDo가 존재 하지 않으면 예외를 발생 시킬려고 나눴습니다...

단순 조회할 경우 findById를 사용하며

업데이트와 삭제시에는 getByid를 사용합니다...ㅎㅎ
getById를 호출하게 되면 getById는 findById를 호출해서 Option을 return 받아서 데이터가 존재 하지 않으면 예외를 발생 시키게끔 코드를 구현했습니다..ㅎㅎ

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/domain/todo/TodoService.java:
```diff
@@ -0,0 +1,11 @@
+public interface TodoService {
+    Todo save(Todo todo);
+    Todo update(String todoId, String title);
+    Todo findById(String id);
+    Todo getById(String id);
```
안녕하세요! `findById` 와 `getById` 둘이 어떻게 차이가 나는지 알고 싶습니다. (어떻게 다르게 쓰려고 하셨는지요!) 각 메서드마다 주석으로 설명을 추가하는 것이 좋을 것 같아요!

goodcodings replied:
앗..... findById에 return 받는 값에 Optional을 감싸지 않았네요 ......
용도는 Repository와 동일합니다...ㅎ

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/AppTest.java:
```diff
@@ -2,8 +2,6 @@
 
 import org.junit.jupiter.api.Test;
 
-import static org.junit.jupiter.api.Assertions.assertNotNull;
```
의존성이 사라지면 테스트가 실행되지 못할 것 같아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/common/util/TodoUtils.java:
```diff
@@ -0,0 +1,49 @@
+    public static Map<String, Object> transferStringToJson(String string) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        Map<String, Object> transferStringToMap = new HashMap<String, Object>();
+        try {
+            transferStringToMap = objectMapper.readValue(string, Map.class);
+        } catch (IOException e) {
+            logger.severe(e.getMessage());
+        }
+        return transferStringToMap;
```
`transferStringToMap` 에 `severe` 한 예외가 생겼지만 `transferStringToMap` 을 그대로 리턴하도록 하는 것보단(이렇게 되면 빈 인풋이 오는 것이 정상일 때와 아닌 때를 구분하기 힘들 것 같아요!) 예외를 던지는 것이 어떨까요?

goodcodings replied:
넵!! 감사합니다 ㅎㅎ parsing하는 부분에 실패할 경우 예외를 던지는 코드로 반영하겠습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/common/util/TodoUtils.java:
```diff
@@ -0,0 +1,49 @@
+
+    public static String transferJsonToString(Object json) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        String transferJsonToString = null;
+        try {
+            transferJsonToString = objectMapper.writeValueAsString(json);
+        } catch (IOException e) {
+            logger.severe(e.getMessage());
+        }
+
+        return transferJsonToString;
```
실패시 null 이 리턴되도록 하는 것은 위험할 것 같아요!

goodcodings replied:
넵!! 실패할 경우 예외를 발생시키는 코드로 변경하겠습니다!!

ssisksl77 commented:
안녕하세요! PR이 계속 커지기때문에 무엇이 바뀐지 정확히 파악하기 힘들어요 ㅠㅜㅠㅜ

현재 PR이 `CodeSoom:main` 을 향해있는데요! 

아래 이미지를 참고해서 `main` 브랜치를 개인 브랜치로 바꿔서 머지 후에 변경된 내역만 따로 PR을 올리면 좋을 것 같아요!
![image](https://user-images.githubusercontent.com/15942102/159931667-836aa00d-9a71-4e8b-920e-3ce2f0de2ed7.png)

goodcodings commented:
> 안녕하세요! PR이 계속 커지기때문에 무엇이 바뀐지 정확히 파악하기 힘들어요 ㅠㅜㅠㅜ
> 
> 현재 PR이 `CodeSoom:main` 을 향해있는데요!aa
> 
> 아래 이미지를 참고해서 `main` 브랜치를 개인 브랜치로 바꿔서 머지 후에 변경된 내역만 따로 PR을 올리면 좋을 것 같아요! ![image](https://user-images.githubusercontent.com/15942102/159931667-836aa00d-9a71-4e8b-920e-3ce2f0de2ed7.png)

죄송합니다 ㅠㅠㅠ 
PR을 처음 사용해보는거라..... 익숙하지가 않네요 ㅠㅠ 죄송합니다 ㅠㅠㅠ

ssisksl77 commented:
> 안녕하세요 과제를 수행하면서 다른 분들의 작성하신 코드를 참조하고 있는데 다른분들은 todo id를 시퀀스로 하고 계셔서.... 제가 잘못하고 있는게 아닐까해서 질문을 드립니다... 객체의 자율성을 보장 할려고 객체 내부에서 아이디를 생성 하는 의미로 코딩을 했는데...
> 
> 제가 잘못된 방향으로 과제를 수행하고 있으면 바로 수정하겠습니다...

ID가 꼭 시퀀스일 필요는 없어요!

johngrib commented:
> > 안녕하세요! PR이 계속 커지기때문에 무엇이 바뀐지 정확히 파악하기 힘들어요 ㅠㅜㅠㅜ
> > 현재 PR이 `CodeSoom:main` 을 향해있는데요!aa
> > 아래 이미지를 참고해서 `main` 브랜치를 개인 브랜치로 바꿔서 머지 후에 변경된 내역만 따로 PR을 올리면 좋을 것 같아요! ![image](https://user-images.githubusercontent.com/15942102/159931667-836aa00d-9a71-4e8b-920e-3ce2f0de2ed7.png)
> 
> 죄송합니다 ㅠㅠㅠ PR을 처음 사용해보는거라..... 익숙하지가 않네요 ㅠㅠ 죄송합니다 ㅠㅠㅠ

`Edit` 버튼 클릭하시면 돼요 ㅎㅎ
![image](https://user-images.githubusercontent.com/1855714/160131064-3823946f-666c-47e4-82e6-a8576664238e.png)
